### ISSUE TITLE
Add SWA to PyTorch mainline

### ISSUE THREAD

**Issue by andrewgordonwilson** [pytorch/pytorch#29994]
## 🚀 Feature

Add [SWA](https://github.com/pytorch/contrib/blob/master/torchcontrib/optim/swa.py) to https://github.com/pytorch/pytorch/tree/master/torch/optim. 

## Motivation

Stochastic weight averaging (SWA) is a general purpose optimization procedure that leads to improved generalization with essentially no computational overhead. [A description of SWA is here](https://pytorch.org/blog/stochastic-weight-averaging-in-pytorch/).

SWA was added to PyTorch contrib in July 2018 and announced in April 2019. Since that time, 126 repos are now using the contrib implementation (https://github.com/search?q=torchcontrib+swa&type=Code). The [separate repo](https://github.com/timgaripov/swa) also has 551 stars and 84 forks (it's less clear in this instance how to search for repos using the implementation). [The paper itself](https://arxiv.org/abs/1803.05407) is also building momentum in terms of citations, and has inspired a number of other papers, over an array of applications (low precision training, Bayesian deep learning, semi-supervised learning, RL, vision, etc.). The reception to the work has also been positive. 

SWA is presently still the only method in contrib. I would like to request that SWA be added to the main library. I believe it could benefit a major contingent of users. SWA is general purpose, easy to use, and building impact over a wide range of domains. 

## Pitch

I would like to request that SWA be added to the main library. I believe it could benefit a major contingent of users. SWA is general purpose, easy to use, and building impact over a wide range of domains. Since it is implemented in PyTorch contrib, the addition would be straightforward.

cc @soumith @vincentqb

**Comment by vincentqb** [comment]
In general, we focus on including methods that the community uses as a standard, or else the code maintenance problem balloons up for us. We do show discretion based on what the paper shows as evidence, for example BatchNorm was included within weeks of it's publish date (in Torch). In terms of rejected methods, we've rejected newly minted papers such as Swish (#3260, #3182), [Yellowfin](https://github.com/pytorch/pytorch/issues/1960) and many others that haven't become standardized in the community (like LSTM/Transformer/BatchNorm).

In this case, since at least a year has passed following the addition in contrib, the github repo and the paper have gathered momentum, I do agree with bringing SWA in pytorch mainline. Would you like to open a pull request with the relevant code, algorithm and tests?

**Comment by andrewgordonwilson** [comment]
Great! We've added a pull request: https://github.com/pytorch/pytorch/pull/30559
